AsyncDisplayKit/Details/ASRangeController.mm

```diff
@@ -255,28 +247,49 @@ - (void)_updateVisibleNodeIndexPaths
   }
 
   ASRangeTuningParameters parametersPreload = [_layoutController tuningParametersForRangeMode:rangeMode
-                                                                                      rangeType:ASLayoutRangeTypePreload];
-  if (ASRangeTuningParametersEqualToRangeTuningParameters(parametersPreload, ASRangeTuningParametersZero)) {
-    preloadIndexPaths = visibleIndexPaths;
-  } else {
-    preloadIndexPaths = [_layoutController indexPathsForScrolling:scrollDirection
-                                                          rangeMode:rangeMode
-                                                          rangeType:ASLayoutRangeTypePreload];
-  }
-  
+                                                                                    rangeType:ASLayoutRangeTypePreload];
   ASRangeTuningParameters parametersDisplay = [_layoutController tuningParametersForRangeMode:rangeMode
                                                                                     rangeType:ASLayoutRangeTypeDisplay];
-  if (rangeMode == ASLayoutRangeModeLowMemory) {
-    displayIndexPaths = [NSSet set];
-  } else if (ASRangeTuningParametersEqualToRangeTuningParameters(parametersDisplay, ASRangeTuningParametersZero)) {
-    displayIndexPaths = visibleIndexPaths;
-  } else if (ASRangeTuningParametersEqualToRangeTuningParameters(parametersDisplay, parametersPreload)) {
-    displayIndexPaths = preloadIndexPaths;
+
+  // Preload can express the ultra-low-memory state with 0, 0 returned for its tuningParameters above, and will match Visible.
+  // However, in this rangeMode, Display is not supposed to contain *any* paths -- not even the visible bounds. TuningParameters can't express this.
+  BOOL emptyDisplayRange = (rangeMode == ASLayoutRangeModeLowMemory);
+  BOOL equalDisplayPreload = ASRangeTuningParametersEqualToRangeTuningParameters(parametersDisplay, parametersPreload);
+  BOOL equalDisplayVisible = (ASRangeTuningParametersEqualToRangeTuningParameters(parametersDisplay, ASRangeTuningParametersZero)
+                              && emptyDisplayRange == NO);
+
+  // Check if both Display and Preload are unique. If they are, we load them with a single fetch from the layout controller for performance.
+  BOOL optimizedLoadingOfBothRanges = (equalDisplayPreload == NO && equalDisplayVisible == NO && emptyDisplayRange == NO);
+
+  NSSet<NSIndexPath *> *visibleIndexPaths = [NSSet setWithArray:visibleNodePaths];
+  NSSet<NSIndexPath *> *displayIndexPaths = nil;
+  NSSet<NSIndexPath *> *preloadIndexPaths = nil;
+  
+  if (optimizedLoadingOfBothRanges) {
```

Comment:
Though a bit complex on the surface, I quite like this code, and think it is pretty readable if you parse & trust the meaning of the BOOLs.

This method should be refactored in general, and I could move some of this to another method in ASRangeController, but for now there is not much net complexity added -- just a few more lines to give more descriptive names to the BOOLs.